### ISSUE TITLE
Tagging seems to be broken with latest BitBucket v5.3.1 and YACC v1.17

### ISSUE THREAD

**Issue by gwsu2008** [sford/yet-another-commit-checker#166]
Have anyone seen this issue? The only way to unblock tagging is disable the plugin all together. 

YACC v1.17
BitBucket v5.3.1
git version 2.14.1
Here's a snippet of the exception on the bitbucket logs file.
```
2017-09-11 22:05:21,393 WARN  [threadpool:thread-1] jenkins @74940Ex1325x173445x0 h5x6rg XXX.XXX.XXX.XXX SSH - git-receive-pack '/cd/cd-sandbox.git' c.a.s.i.h.r.DefaultRepositoryHookService [CD/cd-sandbox[168]] Error calling com.isroot.stash.plugin.YaccHook.preUpdate (com.isroot.stash.plugin.yacc:yaccHook)
com.atlassian.bitbucket.ServerException: An error occurred while executing an external process: Requested object from git cat-file --batch, but did not get output
        at com.atlassian.bitbucket.scm.git.command.GitCommandExitHandler.evaluateThrowable(GitCommandExitHandler.java:113)
        at com.atlassian.bitbucket.scm.git.command.GitCommandExitHandler.onCancel(GitCommandExitHandler.java:50)
        at com.atlassian.bitbucket.scm.BaseCommand.callExitHandler(BaseCommand.java:143)
        at com.atlassian.bitbucket.scm.BaseCommand$CommandFuture.internalGet(BaseCommand.java:280)
        at com.atlassian.bitbucket.scm.BaseCommand$CommandFuture.get(BaseCommand.java:244)
        at com.atlassian.bitbucket.scm.BaseCommand.call(BaseCommand.java:83)
        at com.isroot.stash.plugin.YaccServiceImpl.checkAnnotatedTag(YaccServiceImpl.java:173)
        at com.isroot.stash.plugin.YaccServiceImpl.checkRefChange(YaccServiceImpl.java:99)
        at com.isroot.stash.plugin.YaccServiceImpl.checkRefs(YaccServiceImpl.java:85)
        at com.isroot.stash.plugin.YaccServiceImpl.check(YaccServiceImpl.java:61)
        at com.isroot.stash.plugin.YaccHook.preUpdate(YaccHook.java:35)
        at com.isroot.stash.plugin.YaccHook.preUpdate(YaccHook.java:17)
        at com.atlassian.stash.internal.hook.repository.DefaultRepositoryHookService.preUpdate(DefaultRepositoryHookService.java:798)
        at com.atlassian.stash.internal.hook.repository.DefaultRepositoryHookService.preUpdate(DefaultRepositoryHookService.java:411)
        at com.atlassian.stash.internal.hook.DefaultBuiltInHookHandlerFactory.lambda$preReceive$0(DefaultBuiltInHookHandlerFactory.java:36)
        at com.atlassian.stash.internal.scm.git.GitQuarantineHelper.lambda$null$0(GitQuarantineHelper.java:87)
        at com.atlassian.stash.internal.hook.DefaultHookService.doHandleRequest(DefaultHookService.java:303)
        at com.atlassian.stash.internal.hook.DefaultHookService.handleRequest(DefaultHookService.java:289)
        at com.atlassian.stash.internal.hook.DefaultHookService.handleRawRequest(DefaultHookService.java:230)
        at com.atlassian.stash.internal.hook.DefaultHookService$1.lambda$run$0(DefaultHookService.java:197)
        at com.atlassian.stash.internal.concurrent.DefaultTransferableStateManager$StateTransferringRunnable.run(DefaultTransferableStateManager.java:166)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.lang.Thread.run(Thread.java:745)
        ... 26 frames trimmed
Caused by: java.lang.IllegalStateException: Requested object from git cat-file --batch, but did not get output
        at com.atlassian.stash.internal.scm.git.command.catfile.AbstractBatchCatFileHandler.requestObject(AbstractBatchCatFileHandler.java:96)
        at com.atlassian.stash.internal.scm.git.command.catfile.CallbackAnnotatedTagHandler.process(CallbackAnnotatedTagHandler.java:84)
        at com.atlassian.stash.internal.scm.git.command.catfile.AbstractBatchCatFileHandler.process(AbstractBatchCatFileHandler.java:65)
        at com.atlassian.bitbucket.io.AbstractCommandInputOutputHandler.process(AbstractCommandInputOutputHandler.java:42)
        at com.atlassian.bitbucket.scm.SummarizingProcessHandler$DelegatingInputStreamHandler.process(SummarizingProcessHandler.java:184)
        at com.atlassian.bitbucket.scm.SummarizingProcessHandler$DelegatingInputStreamHandler.process(SummarizingProcessHandler.java:166)
        at com.atlassian.bitbucket.scm.SummarizingProcessHandler.processOutput(SummarizingProcessHandler.java:90)
        at com.atlassian.utils.process.ExternalProcessImpl$StdoutHandler.handle(ExternalProcessImpl.java:794)
        at com.atlassian.utils.process.ExternalProcessImpl$HandlerLatchedRunnable.doTask(ExternalProcessImpl.java:731)
        at com.atlassian.utils.process.LatchedRunnable.run(LatchedRunnable.java:158)
        at com.atlassian.utils.process.ExternalProcessImpl$AsynchronousHandlerLatchedRunnable.run(ExternalProcessImpl.java:700)
        at com.atlassian.stash.internal.concurrent.DefaultTransferableStateManager$StateTransferringRunnable.run(DefaultTransferableStateManager.java:166)
        ... 3 common frames omitted
2017-09-11 22:05:21,393 INFO  [threadpool:thread-1] jenkins @74940Ex1325x173445x0 h5x6rg XXX.XXX.XXX.XXX  SSH - git-receive-pack '/cd/cd-sandbox.git' c.a.s.i.h.r.DefaultRepositoryHookService [CD/cd-sandbox[168]] hook 'yaccHook' vetoed the push request
```

**Comment by sford** [comment]
Hi @gwsu2008 ,

Trying to figure out how to reproduce this... What kind of tags are you trying to push? Lightweight or annotated tags? What is the git command you are using to for the push?

**Comment by sford** [comment]
@gwsu2008 It could help me reproduce the problem if you included YACC logs when the push fails.

You can enable YACC logging by using `curl`, something like:
```
curl -u ADMIN_USER -v -X PUT -d "" -H "Content-Type: application/json" https://YOUR_BBS_HOSTNAME/bitbucket/rest/api/latest/logs/logger/com.isroot/debug
```

**Comment by sford** [comment]
In the meantime, if needed, you can downgrade the previous version of YACC here:
https://marketplace.atlassian.com/plugins/com.isroot.stash.plugin.yacc/versions

**Comment by gwsu2008** [comment]
Hi sford,

Here's the git command. I am trying to push annotated tag.
TAG=GSU-1.0.0-123
+ git tag -a GSU-1.0.0-123 -m GSU-1.0.0-123
+ git push origin : GSU-1.0.0-123
remote: Repository hook com.isroot.stash.plugin.yacc:yaccHook failed        
remote: An error occurred while executing an external process: Requested object from git cat-file --batch, but did not get output  

I will enable debug logging and try again.

Thanks

**Comment by gwsu2008** [comment]
Hi sford,

I enabled the debug and try again. See attach debug log file from bitbucket.
[atlassian-bitbucket.txt](https://github.com/sford/yet-another-commit-checker/files/1296722/atlassian-bitbucket.txt)

Thanks

**Comment by gwsu2008** [comment]
Downgrade to previous version 1.16 fixed the issue so the new version 1.17 introduced the bug.

**Comment by sford** [comment]
Hi @gwsu2008,

Yeah, this is most likely an issue with YACC 1.17 or new BBS 5 annotated tag APIs. YACC 1.17's main feature is using BBS 5 APIs for improved performance. In this case, instead of obtaining annotated tag information itself, it instead asks BBS 5 to do it. The exception you are getting is from BBS 5 code.

Thanks for all the info! That is exactly what I was looking for. Unfortunately, I have yet to reproduce the error, so I still don't know exactly what is going on. I am testing with the same YACC, BBS, and git version, same YACC settings, and using the git tag/push commands that you gave.

What kind of environment are you running BBS on (Windows, Linux, etc)? Is your repository using any advanced features that might be a bit non-standard?

**Comment by gwsu2008** [comment]
Hi @sford 

We are running BitBucket server on CentOS Linux. See attached screen capture for 
repository details and hooks enabled for the repo

![hooks](https://user-images.githubusercontent.com/11874389/30361127-16a3905c-980a-11e7-843e-11e28e988087.jpg)
![repo_details](https://user-images.githubusercontent.com/11874389/30361128-18a311ac-980a-11e7-8d2b-8e2f3abba616.jpg)

Thanks

**Comment by bturner** [comment]
@gwsu2008 I'm one of the Bitbucket Server developers. Can you try enabling the "Verify Committer" hook _just briefly_ and see if you can reproduce the failure? (As with YACC, you'll need to push a new tag to trigger the relevant part of the "Verify Committer" checks.) That hook uses the same annotated tag API that YACC 1.17 has been updated to use, which suggests, if it's a problem with Bitbucket Server's code, you should have a very similar failure.

I've reviewed the YACC code in question and it appears to be using our API correctly, so the next step is going to be trying to formalize a reproduction.

**Comment by ata-sql** [comment]
We have the same issue in our environment. YACC 1.17, BB 5.3.1, git 1.12.2.

**Comment by bturner** [comment]
@ata-sql That Git version can't be correct. Git 1.x cannot be used with Bitbucket Server 5.x. You must be using 2.2.0 or newer or Bitbucket Server would fail on startup. Can you recheck your version? Also, can you please check the "Verify Commiter" hook to see if it fails the same way?

**Comment by ata-sql** [comment]
Sorry, misspelled :). We use 2.12.2.

**Comment by mgillespie** [comment]
Same issue here.   We downgraded to YACC1.16  and problem is fixed  (YACC1.17, BBS 5.3.1, GIT 2.14.2 ).

I was nervous, as none of the documentation indicated if BitbucketServer would remove the plugin settings on plugin uninstall.  Thankfully it didn't, so it was just a case of uninstall 1.17 and upload 1.16

**Comment by bturner** [comment]
@mgillespie I'll ask you the same question I've asked the last 2 reporters: Can you enable the built-in "Verify Committer" hook and see if it causes the same issue when pushing a new tag? The method YACC 1.17 is calling that triggers this failure is the same method the "Verify Committer" hook is calling, so "Verify Committer" should exhibit an identical failure. Assuming it does, the bug must be in Bitbucket Server, rather than YACC.

I need someone who's willing to provide more details in order to make any progress on diagnosing the underlying problem. I'm running Git 2.14.1 and I don't see any issues with verifying tags. A zipped up repository which reproduces this issue would really go a long way toward aiding a solution. Something that someone who has the issue has tested that I can simply unzip and push would let me debug into the processing (or at least confirm that _identical data_ doesn't cause the failure for me)

**Comment by mgillespie** [comment]
Sorry, this is a production server, can't afford to have angry developers who can't check code in...   This feels like it should be easy to reproduce.

**Comment by bturner** [comment]
Unfortunately, it's not easy to reproduce. In fact, we have full integration tests (running real git commands in real repositories across over a dozen versions of Git) that show it works.

Is it not possible to create a repository and test this? Even on a production server, that's not going to impact anyone. I'm not asking for someone to modify a busy repository and try this out.

**Comment by ata-sql** [comment]
@bturner I did that test - disabled yacc and enabled Verify Commiter hook. It worked fine:
> 2017-09-19 17:16:51,901 DEBUG [tag-indexer:thread-1]  c.a.b.internal.tag.idx.TagIndexer [repo-name[195]] tag index written in 1 ms
With yacc enabled i got this:
`
2017-09-19 17:15:40,808 WARN  [threadpool:thread-5] my-login @1GW1I83x1035x228821x0 ieeig1 192.168.x.x SSH - git-receive-pack 'repostiory-name.git' c.a.s.i.h.r.DefaultRepositoryHookService [X/repostiory-name[195]] Error calling com.isroot.stash.plugin.YaccHook.preUpdate (com.isroot.stash.plugin.yacc:yaccHook)
com.atlassian.bitbucket.ServerException: An error occurred while executing an external process: Stream closed
	at com.atlassian.bitbucket.scm.git.command.GitCommandExitHandler.evaluateThrowable(GitCommandExitHandler.java:113)
	at com.atlassian.bitbucket.scm.git.command.GitCommandExitHandler.onError(GitCommandExitHandler.java:201)
	at com.atlassian.bitbucket.scm.DefaultCommandExitHandler.onExit(DefaultCommandExitHandler.java:31)
	at com.atlassian.bitbucket.scm.BaseCommand.callExitHandler(BaseCommand.java:146)
	at com.atlassian.bitbucket.scm.BaseCommand$CommandFuture.internalGet(BaseCommand.java:280)
	at com.atlassian.bitbucket.scm.BaseCommand$CommandFuture.get(BaseCommand.java:244)
	at com.atlassian.bitbucket.scm.BaseCommand.call(BaseCommand.java:83)
	at com.isroot.stash.plugin.YaccServiceImpl.checkAnnotatedTag(YaccServiceImpl.java:173)
	at com.isroot.stash.plugin.YaccServiceImpl.checkRefChange(YaccServiceImpl.java:99)
	at com.isroot.stash.plugin.YaccServiceImpl.checkRefs(YaccServiceImpl.java:85)
	at com.isroot.stash.plugin.YaccServiceImpl.check(YaccServiceImpl.java:61)
	at com.isroot.stash.plugin.YaccHook.preUpdate(YaccHook.java:35)
	at com.isroot.stash.plugin.YaccHook.preUpdate(YaccHook.java:17)
	at com.atlassian.stash.internal.hook.repository.DefaultRepositoryHookService.preUpdate(DefaultRepositoryHookService.java:798)
	at com.atlassian.stash.internal.hook.repository.DefaultRepositoryHookService.preUpdate(DefaultRepositoryHookService.java:411)
	at com.atlassian.stash.internal.hook.DefaultBuiltInHookHandlerFactory.lambda$preReceive$0(DefaultBuiltInHookHandlerFactory.java:36)
	at com.atlassian.stash.internal.scm.git.GitQuarantineHelper.lambda$null$0(GitQuarantineHelper.java:87)
	at com.atlassian.stash.internal.hook.DefaultHookService.doHandleRequest(DefaultHookService.java:303)
	at com.atlassian.stash.internal.hook.DefaultHookService.handleRequest(DefaultHookService.java:289)
	at com.atlassian.stash.internal.hook.DefaultHookService.handleRawRequest(DefaultHookService.java:230)
	at com.atlassian.stash.internal.hook.DefaultHookService$1.lambda$run$0(DefaultHookService.java:197)
	at com.atlassian.stash.internal.concurrent.DefaultTransferableStateManager$StateTransferringRunnable.run(DefaultTransferableStateManager.java:166)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.lang.Thread.run(Thread.java:748)
	... 26 frames trimmed
Caused by: java.io.IOException: Stream closed
	at java.lang.ProcessBuilder$NullOutputStream.write(ProcessBuilder.java:433)
	at java.io.OutputStream.write(OutputStream.java:116)
	at java.io.BufferedOutputStream.flushBuffer(BufferedOutputStream.java:82)
	at java.io.BufferedOutputStream.flush(BufferedOutputStream.java:140)
	at java.io.FilterOutputStream.flush(FilterOutputStream.java:140)
	at sun.nio.cs.StreamEncoder.implFlush(StreamEncoder.java:297)
	at sun.nio.cs.StreamEncoder.flush(StreamEncoder.java:141)
	at java.io.OutputStreamWriter.flush(OutputStreamWriter.java:229)
	at java.io.BufferedWriter.flush(BufferedWriter.java:254)
	at com.atlassian.stash.internal.scm.git.command.catfile.AbstractBatchCatFileHandler.requestObject(AbstractBatchCatFileHandler.java:92)
	at com.atlassian.stash.internal.scm.git.command.catfile.CallbackAnnotatedTagHandler.process(CallbackAnnotatedTagHandler.java:84)
	at com.atlassian.stash.internal.scm.git.command.catfile.AbstractBatchCatFileHandler.process(AbstractBatchCatFileHandler.java:65)
	at com.atlassian.bitbucket.io.AbstractCommandInputOutputHandler.process(AbstractCommandInputOutputHandler.java:42)
	at com.atlassian.bitbucket.scm.SummarizingProcessHandler$DelegatingInputStreamHandler.process(SummarizingProcessHandler.java:184)
	at com.atlassian.bitbucket.scm.SummarizingProcessHandler$DelegatingInputStreamHandler.process(SummarizingProcessHandler.java:166)
	at com.atlassian.bitbucket.scm.SummarizingProcessHandler.processOutput(SummarizingProcessHandler.java:90)
	at com.atlassian.utils.process.ExternalProcessImpl$StdoutHandler.handle(ExternalProcessImpl.java:794)
	at com.atlassian.utils.process.ExternalProcessImpl$HandlerLatchedRunnable.doTask(ExternalProcessImpl.java:731)
	at com.atlassian.utils.process.LatchedRunnable.run(LatchedRunnable.java:158)
	at com.atlassian.utils.process.ExternalProcessImpl$AsynchronousHandlerLatchedRunnable.run(ExternalProcessImpl.java:700)
	at com.atlassian.stash.internal.concurrent.DefaultTransferableStateManager$StateTransferringRunnable.run(DefaultTransferableStateManager.java:166)
	... 3 common frames omitted
	Suppressed: java.io.IOException: Stream closed
		at java.lang.ProcessBuilder$NullOutputStream.write(ProcessBuilder.java:433)
		at java.io.OutputStream.write(OutputStream.java:116)
		at java.io.BufferedOutputStream.flushBuffer(BufferedOutputStream.java:82)
		at java.io.BufferedOutputStream.flush(BufferedOutputStream.java:140)
		at java.io.FilterOutputStream.flush(FilterOutputStream.java:140)
		at java.io.FilterOutputStream.close(FilterOutputStream.java:158)
		at sun.nio.cs.StreamEncoder.implClose(StreamEncoder.java:320)
		at sun.nio.cs.StreamEncoder.close(StreamEncoder.java:149)
		at java.io.OutputStreamWriter.close(OutputStreamWriter.java:233)
		at java.io.BufferedWriter.close(BufferedWriter.java:266)
		at com.atlassian.stash.internal.scm.git.command.catfile.AbstractBatchCatFileHandler.process(AbstractBatchCatFileHandler.java:67)
		... 12 common frames omitted
		Suppressed: java.io.IOException: Stream closed
			at java.lang.ProcessBuilder$NullOutputStream.write(ProcessBuilder.java:433)
			at java.io.OutputStream.write(OutputStream.java:116)
			at java.io.BufferedOutputStream.flushBuffer(BufferedOutputStream.java:82)
			at java.io.BufferedOutputStream.flush(BufferedOutputStream.java:140)
			at java.io.FilterOutputStream.close(FilterOutputStream.java:158)
			at java.io.FilterOutputStream.close(FilterOutputStream.java:159)
			... 17 common frames omitted
`

**Comment by sford** [comment]
@ata-sql Thanks for spot checking Verify Committer hook, that is helpful. Unfortunately that points to a potential problem in YACC (assuming Verify Committer is using the same APIs as YACC), and I haven't been able to reproduce this locally. It is obvious something is going on, but difficult to figure out because I haven't been able to reproduce it.

@bturner Could Verify Committer hook be doing something slightly different than YACC? (or, is it open source? ~ maybe I can spot check the code)

**Comment by bturner** [comment]
@sford Bitbucket Server's full source is available to anyone with a license, including $10 starter licenses.

The relevant code is:
```java
    @Nonnull
    @Override
    public RepositoryHookResult preUpdate(@Nonnull PreRepositoryHookContext context,
                                          @Nonnull RepositoryHookRequest request) {
        if (IGNORED_TRIGGERS.contains(request.getTrigger())) {
            return RepositoryHookResult.accepted();
        }
        ApplicationUser user = authenticationContext.getCurrentUser();
        if (user == null) {
            // anonymous pushes should never happen, but in the spirit of the hook, we should block them
            return RepositoryHookResult.rejected(
                    i18nService.getMessage("bitbucket.plugins.hooks.verifycommitter.summary"),
                    i18nService.getMessage("bitbucket.plugins.hooks.verifycommitter.anonymousdetails"));
        }
        if (user.getType() == UserType.SERVICE) {
            log.trace("[{}] Not verifying committer for service user '{}'", request.getRepository(), user.getSlug());
            return RepositoryHookResult.accepted();
        }

        List<RefChange> relevantRefChanges = request.getRefChanges().stream()
                .filter(this::canIntroduceCommits)
                .collect(Collectors.toList());

        if (!relevantRefChanges.isEmpty()) {
            RepositoryHookResult result = verifyAnnotatedTags(request.getRepository(), relevantRefChanges);
            if (result.isRejected()) {
                return result;
            }
            context.registerCommitCallback(new CommitterVerifyingCommitCallback(i18nService, user), ADDED_TO_REPOSITORY);
        }

        return RepositoryHookResult.accepted();
    }

    private boolean canIntroduceCommits(RefChange refChange) {
        RefChangeType type = refChange.getType();
        return type == RefChangeType.ADD || type == RefChangeType.UPDATE;
    }

    private RepositoryHookResult verifyAnnotatedTags(Repository repository, List<RefChange> relevantRefChanges) {
        if (!GitScm.ID.equals(repository.getScmId())) {
            return RepositoryHookResult.accepted();
        }

        Set<String> tagsToCheck = relevantRefChanges.stream()
                .filter(refChange -> refChange.getRef().getType() == StandardRefType.TAG)
                .map(RefChange::getToHash)
                .collect(toSet());

        if (tagsToCheck.isEmpty()) {
            // no tags to verify
            return RepositoryHookResult.accepted();
        }

        TaggerVerifyingCallback callback = new TaggerVerifyingCallback(repository);
        GitResolveAnnotatedTagsCommandParameters params = new GitResolveAnnotatedTagsCommandParameters.Builder()
                .tagIds(tagsToCheck)
                .maxMessageLength(0)
                .build();
        gitRefCommandFactory.resolveAnnotatedTags(repository, params, callback).call();
        return callback.getResult();
    }
```

So there are 4 differences I can see, when compared to YACC's code:
* We check both `ADD` and `UPDATE` `RefChange`s that affect tags
  * This code isn't shown, but before we call the `verifyAnnotatedTags` method I've shown we first filter the `RefChange`s to only those which can introduce commits, and that's any `ADD` or `UPDATE`
* We explicitly restrict this to Git repositories only
* We only call `resolveAnnotatedTags` _once_, regardless of the number of tags that get pushed, where YACC calls it individually for every tag
* We suppress any tag message that might be present

It looks like YACC could probably be updated to check the SCM and to suppress tag messages very simply. It doesn't seem like it would be very difficult to check all tags at once, either (though it's definitely a bigger change).

Conceptually, though, it doesn't seem like _any_ of those differences would account for why YACC's request fails and "Verify Committer" succeeds.

One thing I've noticed is that, at least so far, I haven't seen anyone comment who's running a version of Git <2.11.0. That means everyone posting here has the "quarantined pushes" changes in Git 2.11, which significantly change how objects are handled during `pre-receive` processing. Unfortunately, I doubt anyone would be willing to downgrade to Git 2.10.x to double-check if the issue goes away (and I _completely_ understand why they wouldn't, to be clear).

**Comment by aslane** [comment]
We also experienced problems with the YACC plugin after upgrading to BitBucket 5.3.2 (from 4.11.2) and upgrading the YACC plugin. We are using git version 2.8.4 so older than the one being asked about. As we are on a production system, we temporarily disabled the YACC plugin to ensure we get our tags pushed with a downgrade to 1.16 planned but the relevant part of the Jenkins log is shown below, should that be of interest.

[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 02:20 min
[INFO] Finished at: 2017-09-26T13:41:34+13:00
[INFO] Final Memory: 104M/1169M
[INFO] ------------------------------------------------------------------------
channel stopped
 > git tag -l fmit-enterprise-2.12.52 # timeout=10
 > git tag -a -f -m Release fmit-enterprise-2.12.52 fmit-enterprise-2.12.52 # timeout=10
Pushing tag fmit-enterprise-2.12.52 to repo origin
 > git --version # timeout=10
using .gitcredentials to set credentials
 > git config --local credential.username jenkins # timeout=10
 > git config --local credential.helper store --file=/tmp/git9021121097237866022.credentials # timeout=10
Setting http proxy: proxy.fm.internal:3128
 > git -c core.askpass=true push https://git.fostermoore.com/scm/fmit-enterprise/fmit-enterprise.git fmit-enterprise-2.12.52 -f
 > git config --local --remove-section credential # timeout=10
ERROR: Failed to push tag fmit-enterprise-2.12.52 to origin
hudson.plugins.git.GitException: Command "git -c core.askpass=true push https://git.fostermoore.com/scm/fmit-enterprise/fmit-enterprise.git fmit-enterprise-2.12.52 -f" returned status code 1:
stdout: 
stderr: remote: Repository hook com.isroot.stash.plugin.yacc:yaccHook failed[K
remote: An error occurred while executing an external process: Requested object from git cat-file --batch, but did not get output[K
To https://git.fostermoore.com/scm/fmit-enterprise/fmit-enterprise.git
 ! [remote rejected] fmit-enterprise-2.12.52 -> fmit-enterprise-2.12.52 (pre-receive hook declined)
error: failed to push some refs to 'https://git.fostermoore.com/scm/fmit-enterprise/fmit-enterprise.git'

	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandIn(CliGitAPIImpl.java:1710)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandWithCredentials(CliGitAPIImpl.java:1454)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.access$300(CliGitAPIImpl.java:63)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl$8.execute(CliGitAPIImpl.java:1786)
	at hudson.plugins.git.GitPublisher.perform(GitPublisher.java:306)
	at hudson.tasks.BuildStepMonitor$3.perform(BuildStepMonitor.java:45)
	at hudson.model.AbstractBuild$AbstractBuildExecution.perform(AbstractBuild.java:779)
	at hudson.model.AbstractBuild$AbstractBuildExecution.performAllBuildSteps(AbstractBuild.java:726)
	at hudson.maven.MavenModuleSetBuild$MavenModuleSetBuildExecution.post2(MavenModuleSetBuild.java:1053)
	at hudson.model.AbstractBuild$AbstractBuildExecution.post(AbstractBuild.java:671)
	at hudson.model.Run.execute(Run.java:1769)
	at hudson.maven.MavenModuleSetBuild.run(MavenModuleSetBuild.java:537)
	at hudson.model.ResourceController.execute(ResourceController.java:98)
	at hudson.model.Executor.run(Executor.java:374)
Build step 'Git Publisher' marked build as failure

**Comment by pgarciafc** [comment]
Also affected by the issue. Using Bitbucket 5.4.0+Git 2.12.2+YACC 1.17.

Downgrade to 1.16 fixed the issue.

**Comment by ata-sql** [comment]
Another issue: BB 5.4, Git 2.14.1, yacc 1.17:
`2017-09-29 11:30:25,201 DEBUG [threadpool:thread-3] $USER @1CUGB0x690x78630x0 1d56akv 192.168.100.48 SSH - git-receive-pack 'repo.git' c.i.stash.plugin.YaccServiceImpl YaccHook preUpdate, registering commit callback.  settings=com.atlassian.stash.internal.setting.MapSettings@33148143
2017-09-29 11:30:25,201 DEBUG [threadpool:thread-3] $USER @1CUGB0x690x78630x0 1d56akv 192.168.100.48 SSH - git-receive-pack 'repo.git' c.i.stash.plugin.YaccServiceImpl refChange: ref=SimpleMinimalRef{id=refs/heads/branch/name, displayId=branch/name} refType=BRANCH type=ADD fromHash=0000000000000000000000000000000000000000 toHash=9b4bb51dccf9d3c043b61aedcab8eb91c2457187
2017-09-29 11:30:25,262 DEBUG [io-pump:thread-141] $USER @1CUGB0x690x78630x0 1d56akv 192.168.100.48 SSH - git-receive-pack 'repo.git' c.i.s.plugin.YaccHookCommitCallback yacc commit callback, ref=refs/heads/branch/name commit=9b4bb51dccf9d3c043b61aedcab8eb91c2457187
2017-09-29 11:30:25,262 DEBUG [io-pump:thread-141] $USER @1CUGB0x690x78630x0 1d56akv 192.168.100.48 SSH - git-receive-pack 'repo.git' c.i.stash.plugin.YaccServiceImpl checking commit id=9b4bb51dccf9d3c043b61aedcab8eb91c2457187 name=USER NAME email=user@domain message=SCOV-2353 [XXXX] bla bla bla
SCOV-2355 [XXXX] bla bla bla
SCOV-2356 [XXXX] bla bla bla
SCOV-2377 [XXXX] bla bla bla
SCOV-1941 [XXXX] bla bla bla
SCOV-2360 [XXXX] bla bla bla
SCOV-2354 [XXXX] bla bla bla
SCOV-2370 [XXXX] bla bla bla
SCOV-1941 [XXXX] bla bla bla
SCOV-2404 [XXXX] bla bla bla
SCOV-2359 [XXXX] bla bla bla
SCOV-2382 [XXXX] bla bla bla
SCOV-2384 [ALL] bla bla bla
SCOV-2384 [XXXX] bla bla bla
SCOV-2379 [XXXX] bla bla bla
SCOV-1941 [XXXX] bla bla bla
DEVELOP-1015 [XXXX] bla bla bla - amend policy version branchName=XXXXXXXXXXXXXXXXXXXXXXXX
2017-09-29 11:30:25,262 DEBUG [io-pump:thread-141] $USER @1CUGB0x690x78630x0 1d56akv 192.168.100.48 SSH - git-receive-pack 'repo.git' c.i.stash.plugin.YaccServiceImpl requireMatchingAuthorEmail=true authorEmail=user@domain stashEmail=user@domain
2017-09-29 11:30:25,262 DEBUG [io-pump:thread-141] $USER @1CUGB0x690x78630x0 1d56akv 192.168.100.48 SSH - git-receive-pack 'repo.git' c.i.stash.plugin.YaccServiceImpl requireMatchingAuthorName=true authorName=USERNAME  USERNAME2 stashName=USERNAME  USERNAME2
2017-09-29 11:30:25,262 DEBUG [io-pump:thread-141] $USER @1CUGB0x690x78630x0 1d56akv 192.168.100.48 SSH - git-receive-pack 'repo.git' c.i.stash.plugin.YaccServiceImpl checking exclude s setting for  $USER: []
2017-09-29 11:30:25,262 DEBUG [io-pump:thread-141] $USER @1CUGB0x690x78630x0 1d56akv 192.168.100.48 SSH - git-receive-pack 'repo.git' c.i.stash.plugin.YaccServiceImpl branch check, excludeBranchRegex= branchName=branch/name
2017-09-29 11:30:25,265 WARN  [io-pump:thread-141] $USER @1CUGB0x690x78630x0 1d56akv 192.168.100.48 SSH - git-receive-pack 'repo.git' org.slf4j.Logger com.isroot.stash.plugin.yacc:yaccHook:onCommitAdded failed: null
2017-09-29 11:30:25,265 DEBUG [io-pump:thread-141] $USER @1CUGB0x690x78630x0 1d56akv 192.168.100.48 SSH - git-receive-pack 'repo.git' c.i.s.plugin.YaccHookCommitCallback callback onEnd
2017-09-29 11:30:25,268 WARN  [threadpool:thread-3]  c.a.s.i.hook.DefaultHookService Hook socket I/O failed before the repository/hook could be identified
com.atlassian.bitbucket.ServerException: An error occurred while executing an external process: detailMessage
        at com.atlassian.bitbucket.scm.git.command.GitCommandExitHandler.evaluateThrowable(GitCommandExitHandler.java:113)
        at com.atlassian.bitbucket.scm.git.command.GitCommandExitHandler.onError(GitCommandExitHandler.java:201)
        at com.atlassian.stash.internal.scm.git.command.revlist.RevCoreExitHandler.onError(RevCoreExitHandler.java:52)
        at com.atlassian.bitbucket.scm.DefaultCommandExitHandler.onExit(DefaultCommandExitHandler.java:31)
        at com.atlassian.bitbucket.scm.BaseCommand.callExitHandler(BaseCommand.java:146)
        at com.atlassian.bitbucket.scm.BaseCommand$CommandFuture.internalGet(BaseCommand.java:280)
        at com.atlassian.bitbucket.scm.BaseCommand$CommandFuture.get(BaseCommand.java:244)
        at com.atlassian.bitbucket.scm.BaseCommand.call(BaseCommand.java:83)
        at com.atlassian.stash.internal.hook.repository.CachingRepositoryHookScmHelper.internalCommits(CachingRepositoryHookScmHelper.java:124)
        at com.atlassian.stash.internal.hook.repository.CachingRepositoryHookScmHelper$HookRequestCache.commits(CachingRepositoryHookScmHelper.java:346)
        at com.atlassian.stash.internal.hook.repository.CachingRepositoryHookScmHelper.commits(CachingRepositoryHookScmHelper.java:71)
        at com.atlassian.stash.internal.hook.repository.DefaultRepositoryHookCallbackInvoker.streamIntroducedCommits(DefaultRepositoryHookCallbackInvoker.java:269)
        at com.atlassian.stash.internal.hook.repository.DefaultRepositoryHookCallbackInvoker.streamCommitDetailsForRefChanges(DefaultRepositoryHookCallbackInvoker.java:203)
        at com.atlassian.stash.internal.hook.repository.DefaultRepositoryHookCallbackInvoker.streamCommitDetails(DefaultRepositoryHookCallbackInvoker.java:131)
        at com.atlassian.stash.internal.hook.repository.DefaultRepositoryHookCallbackInvoker.invokePreCallbacks(DefaultRepositoryHookCallbackInvoker.java:83)
        at com.atlassian.stash.internal.hook.repository.DefaultRepositoryHookService.preUpdate(DefaultRepositoryHookService.java:429)
        at com.atlassian.stash.internal.hook.DefaultBuiltInHookHandlerFactory.lambda$preReceive$0(DefaultBuiltInHookHandlerFactory.java:36)
        at com.atlassian.stash.internal.scm.git.GitQuarantineHelper.lambda$null$0(GitQuarantineHelper.java:87)
        at com.atlassian.stash.internal.hook.DefaultHookService.doHandleRequest(DefaultHookService.java:303)
        at com.atlassian.stash.internal.hook.DefaultHookService.handleRequest(DefaultHookService.java:289)
        at com.atlassian.stash.internal.hook.DefaultHookService.handleRawRequest(DefaultHookService.java:230)
        at com.atlassian.stash.internal.hook.DefaultHookService$1.lambda$run$0(DefaultHookService.java:197)
        at com.atlassian.stash.internal.concurrent.DefaultTransferableStateManager$StateTransferringRunnable.run(DefaultTransferableStateManager.java:166)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.lang.Thread.run(Thread.java:748)
        ... 26 frames trimmed
Caused by: java.lang.NullPointerException: detailMessage
        at java.util.Objects.requireNonNull(Objects.java:228)
        at com.atlassian.bitbucket.hook.repository.SimpleRepositoryHookVeto.<init>(SimpleRepositoryHookVeto.java:20)
        at com.atlassian.bitbucket.hook.repository.RepositoryHookResult$Builder.veto(RepositoryHookResult.java:141)
        at com.atlassian.bitbucket.hook.repository.RepositoryHookResult.rejected(RepositoryHookResult.java:49)
        at com.atlassian.stash.internal.hook.repository.SafeRepositoryHookCommitCallback.addFailureResult(SafeRepositoryHookCommitCallback.java:99)
        at com.atlassian.stash.internal.hook.repository.SafeRepositoryHookCommitCallback.invokeSafely(SafeRepositoryHookCommitCallback.java:148)
        at com.atlassian.stash.internal.hook.repository.SafeRepositoryHookCommitCallback.onCommitAdded(SafeRepositoryHookCommitCallback.java:78)
        at com.atlassian.stash.internal.hook.repository.DefaultRepositoryHookCallbackInvoker$HookCallbackInvokingCallback.onCommit(DefaultRepositoryHookCallbackInvoker.java:316)
        at com.atlassian.stash.internal.hook.repository.CachingRepositoryHookScmHelper$CachingCommitCallback.onCommit(CachingRepositoryHookScmHelper.java:242)
        at com.atlassian.stash.internal.scm.git.command.revlist.CallbackRevListOutputHandler.processReader(CallbackRevListOutputHandler.java:81)
        at com.atlassian.bitbucket.io.LineReaderOutputHandler.process(LineReaderOutputHandler.java:42)
        at com.atlassian.bitbucket.scm.SummarizingProcessHandler$DelegatingInputStreamHandler.process(SummarizingProcessHandler.java:184)
        at com.atlassian.bitbucket.scm.SummarizingProcessHandler$DelegatingInputStreamHandler.process(SummarizingProcessHandler.java:166)
        at com.atlassian.bitbucket.scm.SummarizingProcessHandler.processOutput(SummarizingProcessHandler.java:90)
        at com.atlassian.utils.process.ExternalProcessImpl$StdoutHandler.handle(ExternalProcessImpl.java:794)
        at com.atlassian.utils.process.ExternalProcessImpl$HandlerLatchedRunnable.doTask(ExternalProcessImpl.java:731)
        at com.atlassian.utils.process.LatchedRunnable.run(LatchedRunnable.java:158)
        at com.atlassian.utils.process.ExternalProcessImpl$AsynchronousHandlerLatchedRunnable.run(ExternalProcessImpl.java:700)
        at com.atlassian.stash.internal.concurrent.DefaultTransferableStateManager$StateTransferringRunnable.run(DefaultTransferableStateManager.java:166)
        ... 3 common frames omitted
`

**Comment by dadadan23** [comment]
Also am seeing this issue with Bitbucket 5.2 and YACC 1.17.  Downgraded to YACC 1.16.

**Comment by sford** [comment]
Guys, thanks for your patience. I just identified and fixed the tag bug in b873dd1.

I was unable to reproduce the problem locally using the Atlassian plugin development tool (running BBS locally). As a dart throw, I decided to see what happens if I pushed a tag to a real BBS system and was immediately able to reproduce the problem.

I tracked it down to a null being passed to `resolveAnnotatedTags`. (embarrassing) 

It is interesting to note the behavior difference between a real BBS server, and the local `bitbucket:run` BBS server. I wrote integration tests for YACC that are fired against the BBS development server for the BBS 5 API migration, and a null repository didn't cause any problems there.

**Comment by sford** [comment]
YACC 1.18 released with fix for "Requested object from git cat-file --batch, but did not get output" exceptions.

**Comment by bturner** [comment]
> I tracked it down to a null being passed to `resolveAnnotatedTags`. (embarrassing)

Looks like we omitted a `requireNonNull` check and (through pure chance), everywhere inside the `resolveAnnotatedTags` where the `Repository` is used it's actually `@Nullable`, so the rest of the code never noticed.

I've added explicit `requireNonNull` checks. They'll ship in the next range of point releases we release.

> It is interesting to note the behavior difference between a real BBS server, and the local bitbucket:run BBS server. I wrote integration tests for YACC that are fired against the BBS development server for the BBS 5 API migration, and a null repository didn't cause any problems there.

This is actually due to a Git behavior, believe it or not. When you run your integration tests, Bitbucket Server is running _inside a Git repository_--the YACC repository. That means when Bitbucket Server doesn't apply a working directory to the Git process, the default one ends up being a location where Git's default behavior of walking "up" until it finds a `.git` directory does end up finding the one for your plugin's source code, and the command doesn't fail.

By comparison, when you install Bitbucket Server "for real", it's pretty unlikely it'll end up inside a Git repository, and so the `git cat-file` command fails at startup because it can't find a Git repository.

I think there's actually a way I can harden this inside Bitbucket Server. I'll look into it.
